format/puffin-spec.md

```diff
@@ -0,0 +1,139 @@
+---
+url: puffin-spec
+toc: false
+---
+<!--
+ - Licensed to the Apache Software Foundation (ASF) under one or more
+ - contributor license agreements.  See the NOTICE file distributed with
+ - this work for additional information regarding copyright ownership.
+ - The ASF licenses this file to You under the Apache License, Version 2.0
+ - (the "License"); you may not use this file except in compliance with
+ - the License.  You may obtain a copy of the License at
+ -
+ -   http://www.apache.org/licenses/LICENSE-2.0
+ -
+ - Unless required by applicable law or agreed to in writing, software
+ - distributed under the License is distributed on an "AS IS" BASIS,
+ - WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ - See the License for the specific language governing permissions and
+ - limitations under the License.
+ -->
+
+# Puffin file format
+
+This is a specification for the Puffin, a file format designed to store
+information such as indexes and statistics about data managed in an
+Iceberg table that cannot be stored directly within the Iceberg manifest. A
+Puffin file contains arbitrary pieces of information (here called "blobs"),
+along with metadata necessary to interpret them. The blobs supported by Iceberg
+are documented at [Blob types](#blob-types).
+
+## Format specification
```

Comment:
@rdblue should we have a qualifier in this spec relating to restrictions for how Puffiin data can be used in context of Iceberg?  Specifically, I feel like we should be explicit that Puffin data must not be required in order to read/interpret data in iceberg format (i.e. it can be used for optimizations and annotating data, but must not be required for correct interpretation of table data).

We don't want Puffin to be a vector to extend the table format for representation.

Reply:
@danielcweeks   This is a valid concern.
This PR doesn't introduce any relation between Puffin and Iceberg table format, so there is nothing to worry about yet.
This relation between file format and Iceberg spec is being added in #4945. Can you please take a look there and ensure it doesn't leave any unwanted ambiguities?

Reply:
I'll refer to this from the other PR as it doesn't address this.